job-server/src/main/scala/spark/jobserver/BaseSupervisorActor.scala

```diff
@@ -0,0 +1,260 @@
+      } else {
+        sender ! globalResultActor
+      }
+
+    case GetContext(name) =>
+      if (haveContext(name)) {
+        val actorOpt = getContext(name)._2
+        if (actorOpt.isDefined) {
+          val future = (actorOpt.get ? SparkContextStatus)(getTimeout().seconds)
+          val originator = sender
+          future.collect {
+            case SparkContextAlive => originator ! getContext(name)
+            case SparkContextDead =>
+              logger.info("SparkContext {} is dead", name)
+              self ! StopContext(name)
+              originator ! NoSuchContext
+          }
+        }
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case StopContext(name, retry) =>
+      if (haveContext(name)) {
+        val contextActorOpt = getContext(name)._2
+        if (contextActorOpt.isDefined) {
+          val contextActorRef = contextActorOpt.get
+          logger.info("Shutting down context {}", name)
+          onStopContext(contextActorRef)
+          contextActorRef ! PoisonPill
+          sender ! ContextStopped
+        } else if (retry) {
+          logger.info(
+            "Context {} is not yet started, will try and stop after {} seconds",
+            name,
+            getTimeout()
+          )
+          // send another stopContext, but make the original sender appear as originator
+          context.system.scheduler.scheduleOnce(
+            getTimeout().seconds, self, StopContext(name, false)
+          )(context.dispatcher, sender)
+        } else {
+          logger.error("Context {} never started and was not removed, unexpected!, stopping!")
+          self ! PoisonPill
+        }
+      } else {
+        // if not in retry, its possible something else removed the context
+        // be safe and signal that the context was stopped
+        if (retry) {
+          sender ! NoSuchContext
+        } else {
+          sender ! ContextStopped
+        }
+      }
+
+    case Terminated(actorRef) =>
+      val name: String = actorRef.path.name
+      logger.info("Actor terminated: {}", name)
+      removeContext(name)
+  }
+
+  protected def haveContext(name: String): Boolean = contexts.contains(name)
+  protected def getContext(name: String): SJSContext = contexts(name)
+  protected def removeContext(actorName: String): Unit = {
+    contexts.retain { case (_, (curName, _, _)) => actorName != curName }
+  }
+  protected def listContexts(): Seq[String] = contexts.keys.toSeq
+
+  protected def addContextStarting(name: String, actorName: String) = {
+    contexts(name) = (actorName, None, None)
+  }
+  protected def addContext(name: String, context: SJSContext) = {
+    contexts(name) = context
+  }
+
+  protected def onMemberUp(member: Member): Unit
+  protected def onActorIdentity(memberActors: Any, actorRefOpt: Option[ActorRef]): Unit
+  protected def onStopContext(actor: ActorRef): Unit
+  protected def getTimeout(): Long
+
+  protected def startContext(
```

Comment:
Again same here, prefer params starting on same line as function name, like

    protected def startContext(name: String,
                                                   actorName: String,  
     ...)

(sorry what I mean is name and actorName line up vertically)